sn_networking/src/event.rs

```diff
@@ -255,32 +256,30 @@ impl SwarmDriver {
             } => {
                 debug!(%peer_id, ?connection_id, ?cause, num_established, "ConnectionClosed: {}", endpoint_str(&endpoint));
             }
-            SwarmEvent::OutgoingConnectionError { peer_id, error, .. } => {
-                error!("OutgoingConnectionError to {peer_id:?} - {error:?}");
-                if let Some(peer_id) = peer_id {
-                    // Related errors are: WrongPeerId, ConnectionRefused(TCP), HandshakeTimedOut(QUIC)
-                    let err_string = format!("{error:?}");
-                    let is_wrong_id = err_string.contains("WrongPeerId");
-                    let is_all_connection_failed = if let DialError::Transport(ref errors) = error {
-                        errors.iter().all(|(_, error)| {
-                            let err_string = format!("{error:?}");
-                            err_string.contains("ConnectionRefused")
-                        }) || errors.iter().all(|(_, error)| {
-                            let err_string = format!("{error:?}");
-                            err_string.contains("HandshakeTimedOut")
-                        })
-                    } else {
-                        false
-                    };
-                    if is_wrong_id || is_all_connection_failed {
-                        info!("Detected dead peer {peer_id:?}");
-                        if !self.dead_peers.contains(&peer_id) {
-                            let _ = self.dead_peers.insert(peer_id);
-                            self.send_event(NetworkEvent::PeerRemoved(peer_id));
-                        }
-                        let _ = self.swarm.behaviour_mut().kademlia.remove_peer(&peer_id);
-                        self.log_kbuckets(&peer_id);
+            SwarmEvent::OutgoingConnectionError {
+                peer_id: Some(failed_peer_id),
+                error,
+                connection_id,
+            } => {
+                error!("OutgoingConnectionError to {failed_peer_id:?} on {connection_id:?} - {error:?}");
+
+                let peer_died = match error {
+                    DialError::WrongPeerId { .. } => true,
+                    DialError::Transport(errors) => errors.iter().any(|(_, error)| matches!(error, TransportError::Other(err) if err.kind() == ErrorKind::ConnectionRefused) ),
+                    _ => false,
+                };
+
+                if peer_died {
+                    info!("Detected dead peer {failed_peer_id:?}");
+                    if self.dead_peers.insert(failed_peer_id) {
```

Comment:
In theory, could it happen that a peer is detected as a dead peer twice? E.g. after a few hours.

In that case, inserting will return `false`, but perhaps the peer is still in the routing table?

Reply:
they can come back. but they'd be removed from routing on the first `death`, i think? (and removed from `dead_peers` on rejoin).

(This insert returns if it existed already... so should be functionally the same as before but with one less call?)